File: erykandroidcommon/src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/grouping/GroupedRecyclerAdapter.java
package pl.edu.amu.wmi.erykandroidcommon.recycler.grouping;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import pl.edu.amu.wmi.erykandroidcommon.R;
import pl.edu.amu.wmi.erykandroidcommon.recycler.AbstractViewHolder;

public abstract class GroupedRecyclerAdapter<T extends ListItem, P extends View, U extends AbstractViewHolder<P>> extends RecyclerView.Adapter {

    private List<? extends ListItem> mItems = new ArrayList<>();

    public void setData(List<ListItem> items) {
        this.mItems = items;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    @Override
    public int getItemViewType(int position) {
        return mItems.get(position).getType().type();
    }

    @Override
    public AbstractViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ListItemType.TYPE_HEADER.type()) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_grouping_header, parent, false);
            return new HeaderViewHolder(view);
        } else {
            return createMyViewHolder(parent);
        }
    }

    protected abstract AbstractViewHolder createMyViewHolder(ViewGroup parent);

    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        ListItem item = mItems.get(position);
        if (ListItemType.TYPE_HEADER.type() == item.getType().type()) {
            HeaderViewHolder holder = (HeaderViewHolder) viewHolder;
            holder.setRow();
        } else {
            bindViewHolder((U) viewHolder, (T) item);
        }
    }

    protected abstract void bindViewHolder(U viewHolder, T t);

}

File: erykandroidcommon/src/main/java/pl/edu/amu/wmi/erykandroidcommon/recycler/grouping/GroupedRecyclerAdapter.kt
package pl.edu.amu.wmi.erykandroidcommon.recycler.grouping

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

import pl.edu.amu.wmi.erykandroidcommon.R
import pl.edu.amu.wmi.erykandroidcommon.recycler.AbstractViewHolder

abstract class GroupedRecyclerAdapter<T : ListItem, P : View, U : AbstractViewHolder<P>> : RecyclerView.Adapter<*>() {

    private var mItems: List<ListItem> = ArrayList()

    fun setData(items: List<ListItem>) {
        this.mItems = items
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    override fun getItemViewType(position: Int): Int {
        return mItems[position].type.type()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder<*> {
        if (viewType == ListItemType.TYPE_HEADER.type()) {
            val view = LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_grouping_header, parent, false)
            return HeaderViewHolder(view)
        } else {
            return createMyViewHolder(parent)
        }
    }

    protected abstract fun createMyViewHolder(parent: ViewGroup): AbstractViewHolder<*>

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val item = mItems[position]
        if (ListItemType.TYPE_HEADER.type() == item.type.type()) {
            val holder = viewHolder as HeaderViewHolder
            holder.setRow()
        } else {
            bindViewHolder(viewHolder as U, item as T)
        }
    }

    protected abstract fun bindViewHolder(viewHolder: U, t: T)

}
